lib/sorcery/crypto_providers/argon2.rb

```diff
@@ -0,0 +1,52 @@
+require 'argon2'
+
+module Sorcery
+  module CryptoProviders
+    class Argon2
+      class << self
+        # Setting the option :pepper implements a keyed Argon2 algorithm.
+        attr_accessor :pepper
+        # This is the :cost option for the Argon2 library.
+        # The higher the cost the more secure it is and the longer is take the generate a hash. By default this is 16.
+        # Set this to whatever you want, play around with it to get that perfect balance between
+        # security and performance.
+        def cost
+          @cost ||= 16
+        end
+        attr_writer :cost
+        alias stretches cost
+        alias stretches= cost=
+
+        # Creates a hash for the password passed.rgon2
+        def encrypt(*tokens)
+  	  hasher = ::Argon2::Password.new(m_cost: cost, secret: pepper)
+	  hasher.create(join_tokens(tokens))
+        end
+
+        # Does the hash match the tokens? Uses the same tokens that were used to encrypt.
+        def matches?(hash, *tokens)
+          return false if hash.nil? || hash == {}
+	  ::Argon2::Password.verify_password(join_tokens(tokens), hash, pepper)
+        end
+
+        # This method is used as a flag to tell Sorcery to "resave" the password
+        # upon a successful login, using the new cost
+        def cost_matches?(hash)
+	  true # Not implemented
+        end
```

Comment:
I think your answer is that it's worth implementing. It's not defined in the Argon2 gem, but it's plausible with the Argon2 format. I'll get this done shortly.

Reply:
Weird that the Argon2 gem doesn't support this. I ran into it when looking at the structure of its password class vs BCrypt's. Honestly, this seems like something that should live upstream in the gem, like it does with BCrypt. I'll see if the Argon2 maintainer is open to that change when I get to Argon2 in the rework (should be in the next week, at my current pace).

Reply:
The upstream maintainer is me. The gem was based as much as possible on the upstream C library, and aligning with the way it's used was a bigger priority than acting like a Bcrypy gem. That said, this is a very trivial one liner to implement, I'm just testing it.

Reply:
> The upstream maintainer is me.

Oh... I feel dumb for not noticing that sooner. :sweat_smile: I've been somewhat bullish with my attention while I plunk away on the Sorcery v1 rework.

> The gem was based as much as possible on the upstream C library, and aligning with the way it's used was a bigger priority than acting like a BCrypt gem. That said, this is a very trivial one liner to implement, I'm just testing it.

Would you be okay with deviating slightly from the 'unaltered' design tenant for the ruby wrapper? I feel it would save a lot of downstream developer effort if we exposed some key helpers within the ruby wrapper, rather than forcing those helpers to be downstream.

This would of course still be a git sub-module, and keep all of the functional aspects delegated to the C implementation. I would also be willing to help co-maintain the wrapper, and open a PR to add these changes where we can work together to find a good balance between maintainability and downstream benefits.

Here's my work on the BCrypt implementation last night, it should hopefully give you an idea of the helpers that I'm looking for: [bcrypt.rb](https://github.com/Sorcery/sorcery-rework/blob/e6e78b71226ba862b9ce167e66306eb6f8ba4fd1/sorcery-core/lib/sorcery/crypto_providers/bcrypt.rb)